src/IO/ReadIndirectBufferFromWebServer.cpp

```diff
@@ -0,0 +1,123 @@
+#include "ReadIndirectBufferFromWebServer.h"
+
+#include <common/logger_useful.h>
+#include <Core/Types.h>
+#include <IO/ReadWriteBufferFromHTTP.h>
+#include <IO/ConnectionTimeoutsContext.h>
+#include <thread>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_SEEK_THROUGH_FILE;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+    extern const int NETWORK_ERROR;
+}
+
+
+ReadIndirectBufferFromWebServer::ReadIndirectBufferFromWebServer(const String & url_,
+                                                                 ContextPtr context_,
+                                                                 size_t max_read_tries_,
+                                                                 size_t buf_size_)
+    : BufferWithOwnMemory<SeekableReadBuffer>(buf_size_)
+    , log(&Poco::Logger::get("ReadIndirectBufferFromWebServer"))
+    , context(context_)
+    , url(url_)
+    , buf_size(buf_size_)
+    , max_read_tries(max_read_tries_)
+{
+}
+
+
+std::unique_ptr<ReadBuffer> ReadIndirectBufferFromWebServer::initialize()
+{
+    Poco::URI uri(url);
+
+    ReadWriteBufferFromHTTP::HTTPHeaderEntries headers;
+    headers.emplace_back(std::make_pair("Range", fmt::format("bytes={}-", offset)));
+    LOG_DEBUG(log, "Reading from offset: {}", offset);
+
+    return std::make_unique<ReadWriteBufferFromHTTP>(
+        uri,
+        Poco::Net::HTTPRequest::HTTP_GET,
+        ReadWriteBufferFromHTTP::OutStreamCallback(),
+        ConnectionTimeouts::getHTTPTimeouts(context),
+        0,
+        Poco::Net::HTTPBasicCredentials{},
+        buf_size,
+        headers);
+}
+
+
+bool ReadIndirectBufferFromWebServer::nextImpl()
+{
+    bool next_result = false, successful_read = false;
+
+    if (impl)
+    {
+        impl->position() = position();
+        assert(!impl->hasPendingData());
+    }
+    else
+    {
+        impl = initialize();
+        next_result = impl->hasPendingData();
+    }
+
+    for (size_t try_num = 0; (try_num < max_read_tries) && !next_result; ++try_num)
+    {
+        try
+        {
+            next_result = impl->next();
+            successful_read = true;
+            break;
+        }
+        catch (const Exception & e)
+        {
+            LOG_WARNING(log, "Read attempt {}/{} failed from {}. ({})", try_num, max_read_tries, url, e.message());
```

Comment:
Not important, but we could create a profile event like `WebServerReadRequestsErrors` and increment it here. I think it might be useful